### PR TITLE
from_parquet: use virtual filesystem to preserve partition information when using cache

### PR DESCRIPTION
**Pull request by skshetry**
Alternative proposal to #744, that uses a `ReferenceFileSystem` (which acts as a pointer) to preserve partitioned information.

### CODE REVIEW

**Comment by shcheklein** [comment]
@skshetry so, which PR is better to review? :)

**Comment by skshetry** [comment]
> so, which PR is better to review? :)

I have closed the other PR as I think using a virtual fs is a much cleaner and less fragile solution.

File: src/datachain/lib/arrow.py [hunk 1]
```diff
@@ -190,18 +225,6 @@ def arrow_type_mapper(col_type: pa.DataType, column: str = "") -> type:  # noqa:
     raise TypeError(f"{col_type!r} datatypes not supported, column: {column}")
 
 
-def _nrows_file(file: File, nrows: int) -> str:
```
Comment (skshetry) [inline src/datachain/lib/arrow.py:193]: I have removed this hack. If `nrows` is a significant number, this is likely to be slower. 

We now iterate up to `self.nrows` with the pyarrow dataset.
I tried running `json-csv-reader` example, and the new method is slightly faster.

File: src/datachain/lib/dc.py [hunk 2]
```diff
@@ -557,7 +557,9 @@ def jmespath_to_name(s: str):
                 nrows=nrows,
             )
         }
-        return chain.gen(**signal_dict)  # type: ignore[misc, arg-type]
+        # disable prefetch if nrows is set
+        settings = {"prefetch": 0} if nrows else {}
```
Comment (skshetry) [inline src/datachain/lib/dc.py:561]: disabled `prefetch` when `nrows` is set.

File: src/datachain/lib/dc.py [hunk 3]
```diff
@@ -1976,8 +1981,6 @@ def from_csv(
             else:
                 msg = f"error parsing csv - incompatible output type {type(output)}"
                 raise DatasetPrepareError(chain.name, msg)
-        elif nrows:
-            nrows += 1
```
Comment (skshetry) [inline src/datachain/lib/dc.py:1980]: This was a workaround to make `nrows` works if the data contains headers. This is no longer needed as we count the rows internally.

File: src/datachain/lib/arrow.py [hunk 4]
```diff
@@ -25,7 +27,17 @@
 DATACHAIN_SIGNAL_SCHEMA_PARQUET_KEY = b"DataChain SignalSchema"
 
 
+class ReferenceFileSystem(fsspec.implementations.reference.ReferenceFileSystem):
+    def _open(self, path, mode="rb", *args, **kwargs):
+        # `fsspec`'s `ReferenceFileSystem._open` reads the whole file in-memory.
```
Comment (shcheklein) [inline src/datachain/lib/arrow.py:32]: @skshetry what are the practical implications of this? does it mean we can't handle large files anymore?

File: src/datachain/lib/arrow.py [hunk 5]
```diff
@@ -55,57 +67,80 @@ def __init__(
     def process(self, file: File):
```
[the remark this answers is not in the document]
Comment (shcheklein) [reply src/datachain/lib/arrow.py:67]: thanks!

File: src/datachain/lib/dc.py (hunk 2 above)
Comment (shcheklein) [reply src/datachain/lib/dc.py:561]: should we just let users disable it though?

File: src/datachain/lib/arrow.py (hunk 4 above)
Comment (skshetry) [reply src/datachain/lib/arrow.py:32]: The comment is explaining the rationale why I am implementing `_open` method.
The `fsspec.ReferenceFileSystem._open` reads the whole file in memory and gives you a `io.BytesIO()` object. I have _fixed_ it to return a file object. 

So this should now be able to handle large files.

**Review by dreadatour — APPROVED** [review]
Looks good to me overall 👍 Went through all the changes carefully.
Same time I am not really familiar yet with this part of the codebase 😢

**Review by shcheklein — APPROVED** [review]

File: src/datachain/lib/dc.py (hunk 2 above)
Comment (skshetry) [reply src/datachain/lib/dc.py:561]: The fact that we are doing pre-downloading/pre-fetching is not going to be obvious to the users. So, finding the right settings to disable/enable them is going to be difficult.

Similarly, the default settings with prefetch is going to be suboptimal when you only want to read a few rows. In fact, I'd have to adjust example tests to disable `prefetch` as well, as we use `laion` dataset to read just a couple of rows.

That said, I do not have any strong opinion, though — please let me know what you think.

**Pull request merged**